### PR TITLE
Update ruby default uri to 0.12.2 and bundled uri to 0.10.3 and fix CVE-2023-36617

### PR DESCRIPTION
**Pull request by Redent0r**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
Patch ruby-uri version and fix CVE-2023-36617 (vulnerability not present in rub uri 12.2) https://github.com/hsbt/ruby/commit/ab5421547c5546603c2383085005272ba7754fea

###### Change Log  

- CVE-2023-36617

###### Does this affect the toolchain?  

**NO**

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-2023-36617

###### Test Methodology

- Unified build: https://dev.azure.com/mariner-org/mariner/_build/results?buildId=408103&view=results

### CODE REVIEW

File: SPECS/ruby/ruby.spec [hunk 1]
```diff
@@ -97,6 +97,7 @@ Source4:        rubygems.con
 Source5:        rubygems.prov
 Source6:        rubygems.req
 Source7:        macros.rubygems
+Patch0:         CVE-2023-36617.patch
```
Comment (neha170) [inline SPECS/ruby/ruby.spec:100]: Could you add a note here why this patch was needed, and if needs to be modified when ruby/rubygem-uri is updated in the future, thanks

Comment (Redent0r) [reply SPECS/ruby/ruby.spec]: Added comment to patch and SPEC file. Thanks!

**Review by neha170 — COMMENTED** [review]
From what I understand, the vulnerability is fixed in uri v0.12.2 and above. We are currently at 0.11.1. So you have updated to 0.12.2.
As per the list of gems for ruby 3.1.4, [Standard Gems 3.1.4 (stdgems.org)](https://stdgems.org/3.1.4/), ruby 3.1.4 comes with uri 0.12.1. This also matches with lib/uri/version.rb file. So, the source code for uri in this ruby is v0.12.1. We can change the version to any number we want, but the source code will still be v0.12.1. As per my understanding, this is not really getting rid of the vulnerability in 0.12.1. It is simply updating the version to another number artificially.
I think the right approach would be to either 1) upgrade ruby to a version that comes with higher source code version of uri (0.12.2 or above), or 2) make sure that the vulnerable code is patched in 0.12.1

**Review by neha170 — APPROVED** [review]
Thanks Saul, the new patch looks good. Please merge once it builds cleanly.
For reference, I looked at the changes between 12.1 and 12.2 here https://github.com/ruby/uri/compare/v0-12 They match with your patch, and also resonate with your reasoning above.

**Comment by Redent0r** [comment]
\*sigh\* wait, in 3.1.4 there's https://github.com/ruby/ruby/blob/v3_1_4/lib/bundler/vendor/uri/lib/uri/version.rb#L3 (lib/bundler/vendor/uri/lib/uri/, version 0.10.2) and https://github.com/ruby/ruby/blob/v3_1_4/lib/uri/version.rb#L3 (lib/uri/, version 0.12.1). https://github.com/hsbt/ruby/commit/c7af39ecd8d78d9ffc312dcde2af3466c7bbaf5f patches the bundler version, but based on standard gems for 3.1.4, I think we ought to fix the one under lib/uri, since its version matches the default gem version.
image

To add salt to the wound, there doesn't seem to be a straight commit/patch to 0.12.2 for lib/uri. I *think* what we want it's a mixture of:
* https://github.com/ruby/uri/commit/eaf89cc31619d49e67c64d0b58ea9dc38892d175 (fixes lib/uri/rfc3986_parser.rb)
* https://github.com/ruby/ruby/commit/1eff362492f4a0ca8d6d036c958afa7961e827f0 (fixes lib/uri/rfc2396_parser.rb)
* https://github.com/ruby/ruby/commit/ab5421547c5546603c2383085005272ba7754fea (bumps up the version)

I can make a single patch out of that. These changes also align exactly with https://github.com/ruby/uri/compare/v0-12

**Comment by Redent0r** [comment]
Alrighty, I ended up applying it in both places to be sure. I bumped the lib version from 0.12.1 to 0.12.2, but left the vendored version the same, since [its currently set at 0.10.2 ](https://github.com/ruby/ruby/blob/v3_1_4/lib/bundler/vendor/uri/lib/uri/version.rb#L3). It didn't felt right to change it to 0.12.2 since it may be missing other things. 

BB: https://dev.azure.com/mariner-org/mariner/_build/results?buildId=408552&view=results [Pass]

**Review by neha170 — COMMENTED** [review]
The lib/ changes look good.
I think we can bump up the version to 10.0.3 in bundler. As per the CVE page, the vulnerability is fixed in 10.0.3. Please verify that the patch to update (and fix vulnerability) to 10.0.3 is the same as the one you've used. Thoughts?

**Comment by Redent0r** [comment]
Yep! If we ignore the test changes, it's exactly the same changes that fix the CVE https://github.com/ruby/uri/compare/v0.10.2...v0.10.3, namely https://github.com/ruby/uri/compare/v0.10.2...v0.10.3#diff-6792cbb51ee9c0c764dc71cdbb424a70fc55849b1d470d80688c352aba088a8d and https://github.com/ruby/uri/compare/v0.10.2...v0.10.3#diff-2fff240f8e962830911219fe98478063f84563f498ac32d327f73e5b357e6725

Excellent! I'll bump the bundler version on the patch to 0.10.3

**Comment by Redent0r** [comment]
New BB: https://dev.azure.com/mariner-org/mariner/_build/results?buildId=409089&view=results

**Review by neha170 — APPROVED** [review]
Looks good, thanks for making the changes

**Pull request merged**